#
# This patch is a fix for a number of issues attributed to subfs, which wasn't
# the cause, just the catalyst.
#
# The race is between umount and unlink, when the name gets released before
# the iput completes, which allows an umount to succeed. Since the iput
# has already removed the inode from the inode list, and is performing
# the ->delete_inode fs method without the inode_lock, generic_shutdown_super
# happily shuts down the superblock, leaving the iput half executed when all
# its important data structures disappear. Oopsen ensue.
#
# Fix by Al Viro <viro@parcelfarce.linux.theplanet.co.uk>
#
--- linux/fs/namei.c	Mon Sep 13 01:32:00 2004
+++ linux/fs/namei.c.fix	Sat Oct  2 05:48:21 2004
@@ -1825,13 +1825,12 @@
 		dput(dentry);
 	}
 	up(&nd.dentry->d_inode->i_sem);
+	if (inode)
+		iput(inode);	/* truncate the inode here */
 exit1:
 	path_release(&nd);
 exit:
 	putname(name);
-
-	if (inode)
-		iput(inode);	/* truncate the inode here */
 	return error;
 
 slashes:
